pc2ccs/pc2v9
# I999 Add configurable execute grace periods for sandbox runs

### Description of what the PR does

1. Add 2 new values to **ContestInformation**: `sandboxGraceTimeSecs` and `sandboxInteractiveGraceMultiplier`.  These integer values may be configured using the `system.pc2.yaml` at contest load time, or the GUI **Settings** tab after a contest is configured.  Default values are: 1 and 3 respectively.
2. Change **Executable** to pad execution timers to wait additional time (grace time) for sandbox and interactive sandbox problems, where the sandbox will monitor the actual CPU time of the submission.  The timers in PC2, in this case, are for **Wall** time to protect against sleeping/wall time consuming processes.
3. Cleaned up some comments that were wrong or had errors in them
4. Rearranged some code to initialize the timer delays earlier on since they are used in 2 places.
5. CI: While testing the new sandbox time grace periods, a couple of bugs were found in the **SubmitSampleRunsPane** (which is used to submit the judge's sample solutions by an administrator.    The following were fixed/added to support **SubmitSampleRuns**:

    - Add alphanumeric sorting instead of pure numeric sorting for some columns since some columns may have both text and numbers in them.
    - Synchronize the add runs since we appear to get 2 runAdded notifications for each run from different worker threads.
    - Fixed bug in Filter where if a custom filter group is used alone, it would not allow you to enable filtering.

Basically, these new configuration parameters replace constants that were used in Executable previously.

This PR is one of those cases where there appear to be lots of changes, but again, Eclipse has made many of them related to spaces or other syntax fixes in `ContestInformation.java`.  Also, there were new modules added and one renamed, so they appear as a lot of differences as well.

### Issue which the PR addresses
Fixes #999 

### Environment in which the PR was developed (OS,IDE, Java version, etc.)
Windows 11/Ubuntu 22.04.1 (tested with both)
java version "1.8.0_321"
Java(TM) SE Runtime Environment (build 1.8.0_321-b07)
Java HotSpot(TM) 64-Bit Server VM (build 25.321-b07, mixed mode)

### Precise steps for _testing_ the PR (i.e., how to demonstrate that it works correctly)
1) Choose a CDP for testing
2) Edit the **system.pc2.yaml** (or **system.yaml**, if there is no **system.pc2.yaml**) and add the following properties:
```
     sandbox-grace-time-secs: 2
     sandbox-interactive-time-multiplier: 4

```
3) Load the contest
4) Start the pc2server an administrator client.
5) Go to the "**Configure Contest->Settings**" tab.
6) Notice that the values for the **Sandbox Grace Time added** and **Interactive problem time multiplier** reflect the values added to the system **yaml** file:
![image](https://github.com/user-attachments/assets/f58cb03f-1229-42f7-98e6-c583aa15823d)

7) Using the GUI, change the values to 1 and 3 respectively.  (You can click the ? (what's this) buttons next to each value for a detailed description of what they do).
8) Shutdown the admin and the server
9) Restart the server and admin
10) Go to the "**Configure Contest->Settings**" tab and notice the values that were modified by the GUI have been saved.
![image](https://github.com/user-attachments/assets/eaebf92e-3194-4868-a6d1-35778013c9a4)
